### PR TITLE
Added C-API hash function, used them in Python interface

### LINKED ISSUE #483: All the Sycl* extension types should define a __hash__ function.
Currently, the `Sycl*` Cython extension types can not be hashed as we do not define the `__hash__` function. Due to the lack of the `__hash__` function dpctl types cannot be stored as dictionary keys.

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
Closes #483 

Superseded #489 (and reuses some changes from there).

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
@diptorupd CI is failing in testing-wheels, and I am not entirely sure why. I was able to to `conda build` locally on Linux.

**Comment by diptorupd** [comment]
@oleksandr-pavlyk I suggest drop 41bd9ab from the PR.

**Comment by oleksandr-pavlyk** [comment]
> @oleksandr-pavlyk I suggest drop [41bd9ab](https://github.com/IntelPython/dpctl/commit/41bd9abaf87ec381e9b24423a087d85b93d4976e) from the PR. It unnecessarily will introduce an incomplete fix into our master.

I will do a gentle history rewriting once the CI winks an approval.

File: dpctl-capi/include/dpctl_sycl_context_interface.h [hunk 1]
```diff
@@ -161,4 +161,15 @@ DPCTLContext_GetBackend(__dpctl_keep const DPCTLSyclContextRef CtxRef);
 DPCTL_API
 void DPCTLContext_Delete(__dpctl_take DPCTLSyclContextRef CtxRef);
 
+/*!
+ * @brief Wrapper over
+ * std::hash<sycl::context>'s operator()
+ *
+ * @param    CtxRef        The DPCTLSyclContextRef pointer.
+ * @return   Hash value of the underlying sycl::context instance.
```
Comment (diptorupd) [inline dpctl-capi/include/dpctl_sycl_context_interface.h:169]: Please wrap the `sycl::context` in double back-ticks? I have started adding double back-tics in docstrings for SYCL API names as it renders better in Sphinx.

File: dpctl-capi/include/dpctl_sycl_context_interface.h [hunk 2]
```diff
@@ -161,4 +161,15 @@ DPCTLContext_GetBackend(__dpctl_keep const DPCTLSyclContextRef CtxRef);
 DPCTL_API
 void DPCTLContext_Delete(__dpctl_take DPCTLSyclContextRef CtxRef);
 
+/*!
+ * @brief Wrapper over
```
Comment (diptorupd) [inline dpctl-capi/include/dpctl_sycl_context_interface.h:165]: Why a line break here?

File: dpctl-capi/include/dpctl_sycl_queue_interface.h [hunk 3]
```diff
@@ -331,4 +331,14 @@ void DPCTLQueue_MemAdvise(__dpctl_keep DPCTLSyclQueueRef QRef,
 DPCTL_API
 bool DPCTLQueue_IsInOrder(__dpctl_keep const DPCTLSyclQueueRef QRef);
 
+/*!
+ * @brief C-API wrapper for std::hash<sycl::queue>'s operator().
+ *
+ * @param    QRef         An opaque pointer to the sycl queue.
```
Comment (diptorupd) [inline dpctl-capi/include/dpctl_sycl_queue_interface.h:337]: should be `sycl::queue` with the backticks.

File: dpctl/_sycl_context.pyx [hunk 4]
```diff
@@ -335,6 +336,13 @@ cdef class SyclContext(_SyclContext):
         else:
             return False
 
+    def __hash__(self):
+        """
+        Returns hash value corresponding to ``self._ctxt_ref``.
```
Comment (diptorupd) [inline dpctl/_sycl_context.pyx:341]: Most probably the comment was introduced by me, but how about: "Returns a has value by hashing the underlying SYCL context object." Similar changes elsewhere too.

File: dpctl/tests/test_sycl_context.py [hunk 5]
```diff
@@ -157,3 +159,12 @@ def test_context_multi_device():
     shmem_1 = dpmem.MemoryUSMShared(256, queue=q1)
     shmem_2 = dpmem.MemoryUSMDevice(256, queue=q2)
     shmem_2.copy_from_device(shmem_1)
+
+
+def test_hashing_of_context():
+    """
+    Test that a SyclContext object can be used as a dictionary key.
```
Comment (diptorupd) [inline dpctl/tests/test_sycl_context.py:166]: s/SyclContext/:class:`dpctl.SyclContext`/. Although we do not generate docs for tests so the point is moot.

**Review by diptorupd — APPROVED** [review]
Minor documentation fixes. Otherwise, LGTM.

File: dpctl-capi/include/dpctl_sycl_context_interface.h (hunk 2 above)
Comment (oleksandr-pavlyk) [reply dpctl-capi/include/dpctl_sycl_context_interface.h]: fixed.

**Pull request merged**